docs: docs for dwh cost estimates and savings (#3554)
TobikoData/sqlmesh: docs/cloud/features/costs_savings.md

If you use one of the supported pricing plans, in order to configure Tobiko Cloud to show cost information you will need to visit Settings.

![Image highlighting location of the Settings link in the left site navigation](./costs_savings/costs-navigation.png)

On the General settings page (1), select your pricing plan (2), enter your costs, and then save (3).

![Annotated image showing locations of the general settings link, pricing plan form fields, and save button](./costs_savings/costs-steps.png)

## Where to find cost and savings estimates information

Estimated costs and savings are displayed on the homepage, production environment page, runs and plans pages, and individual model pages.

Cost information on each page will look similar to the following:

![Example of costs and savings data as seen on the Tobiko Cloud homepage](./costs_savings/costs-example.png)

### Savings Categories

When calculating your data warehouse costs, we will calculate the amount you saved by using Tobiko as well! Tobiko Cloud also comes with additional change categorization capabilities, such as advanced column-level impact analysis.

Cost savings are broken up into three main categories:

> - **Prevented Reruns**: If SQLMesh has already run an execution for a change in one environment, we won't need to rerun it in another environment (such as when a model change is backfilled on a development environment and then applied to prod)

Reviewer: It feels the virtual environments bullet is describing this same functionality

Author: I think the two are indeed similar so it makes sense that it feels the same, but maybe there's a better way to differentiate? - I can double check with @crericha that this language is correct

Reviewer 2: I think the rewording you did, after Trey made his comment, addressed the concern and the two classifications sound distinct.

# Data Warehouse Costs and Savings with Tobiko

> Cloud data warehouses can be expensive, and finance departments are paying attention. When they come asking about your cloud budget, the best preparation is a granular understanding of how your project's components correspond to those costs.

Reviewer 3: ```suggestion
Cloud data warehouses can be expensive, and finance departments are paying attention. When they come asking about your cloud budget, the best preparation is a granular understanding of how your project's data pipelines map to those costs.
```

Cloud warehouse costs are determined by how much computation occurs. Clouds report computation/cost for each executed query, but it is challenging to align every query with a specific model in your project.

Tobiko Cloud solves this problem for you. It tracks data warehouse cost estimates per model for BigQuery and Snowflake projects that use a supported pricing plan.

> This granular cost information allows you to directly explain/justify your cloud warehouse expenditures. It also uncovers the models that could most benefit from efficiency improvements.

Reviewer 3: ```suggestion
This granular cost information allows you to directly explain/justify your cloud warehouse spend. It also uncovers the models that could most benefit from efficiency improvements.
```

Beyond tracking your warehouse costs, Tobiko Cloud estimates how much money it saved you! Its advanced column-level impact analysis saves you even more money than open-source SQLMesh.

## Supported Data Warehouse Pricing Plans

Tobiko Cloud supports these cloud pricing plans:

- BigQuery [On Demand](https://cloud.google.com/bigquery/pricing#on_demand_pricing)
- Snowflake [Credits](https://docs.snowflake.com/en/user-guide/cost-understanding-compute#label-what-are-credits)

## Data Warehouse Cost Configuration

Configure Tobiko Cloud to show cost information by navigating to the Settings page.

![Image highlighting location of the Settings link in the left site navigation](./costs_savings/costs-navigation.png)

> On the General settings page (1), select your pricing plan (2), enter your costs, and then save (3).

Reviewer 3: What are default pricing plans? Include configs in writing or screenshots for those

Author: I think I need a little clarification on this one - are you asking what the default values are for the pricing plans, or just what the form looks like? I could update the `costs-steps.png` file to include the full form with either bigquery or snowflake selected that would show the "cost per TB" or "cost per credit" field as well, or is there something else you are looking for?

Reviewer 3: "full form with either bigquery or snowflake selected that would show the "cost per TB" or "cost per credit" field as well"

Can you include screenshots of both and explicitly write out the defaults? Ex: BigQuery charges $6.25 per TB processed

<img width="334" alt="image" src="https://github.com/user-attachments/assets/a8b4f7f5-2987-480b-9ef4-27f75eb0b45c" />

Author: Is this along the lines of what you were thinking?

<img width="879" alt="image" src="https://github.com/user-attachments/assets/3a3b3b1b-066b-4777-a2fa-415f6cecc279" />

Reviewer 2: I worry the code and documentation will get out of sync. It will be hard to remember the docs need to be updated if the defaults change.

Reviewer 3: Yes, these are the screenshots I was looking for. And @crericha fair point, so @caiters can you add a note along the lines of "These are the current cost defaults at the time of setup and update in Tobiko Cloud as warehouse cost defaults change"

![Annotated image showing locations of the general settings link, pricing plan form fields, and save button](./costs_savings/costs-steps.png)

## Where to find cost and savings estimates information

Estimated costs and savings are displayed on the Tobiko Cloud homepage, production environment page, runs and plans pages, and individual model pages.

Cost information on each page will look similar to:

![Example of costs and savings data as seen on the Tobiko Cloud homepage](./costs_savings/costs-example.png)

## Savings Categories

When calculating your data warehouse costs, Tobiko Cloud also estimates how much money it saved you!

> Cost savings are broken up into three main categories:

Reviewer 3: do we break it down visually or is this math only behind the scenes?

Author: we show them aggregated on the main view, and then if you look at the little info icon you'll get a tooltip with the breakdown!

See image from the cost savings PR here:
![image](https://github.com/user-attachments/assets/84a0addd-fc4e-41ac-9679-2bc4ebef2374)

Reviewer 3: Replace the costs-example pic with this one you replied in thread. This is much better for the tooltip.

- **Prevented Reruns**: If SQLMesh has already run an execution for a change in one environment, we won't need to rerun it in another environment (such as when a model change is backfilled on a development environment and then applied to prod).
- **Unaffected Downstream**: Because SQLMesh understands SQL, we know if a downstream model is not affected by an upstream change and can avoid re-execution.
- **Virtual Environments**: Using the Virtual data environments feature means new environments can be created without backfilling models within the new environment.

### Where to find cost savings information

> Cost savings are shown in most of the places data warehouse costs are displayed. You can find how much Tobiko has saved you by viewing the homepage, production environment page, or individual model pages.

Reviewer 3: ```suggestion
Cost savings are shown in most of the places data warehouse costs are displayed. You can find how much Tobiko Cloud has saved you by viewing the homepage, production environment page, or individual model pages.
```

## Data Warehouse Cost Configuration

Configure Tobiko Cloud to show cost information by navigating to the Settings page.

![Image highlighting location of the Settings link in the left site navigation](./costs_savings/costs-navigation.png)

On the General settings page (1), select your pricing plan (2), enter your costs, and then save (3).

![Annotated image showing locations of the general settings link, pricing plan form fields, and save button](./costs_savings/costs-steps.png)

For the BigQuery On Demand plan, we supply BigQuery's default cost per terabyte processed of $6.25\*:
![Costs form for BigQuery On-Demand](./costs_savings/costs-bigquery-on-demand.png)

For the Snowflake Credits plan, we supply Snowflake's default cost per credit of $3.00\*:
![Costs form for Snowflake Credits](./costs_savings/costs-snowflake-credits.png)

If your organization has negotiated a different cost, please enter that cost into the "Cost per" field instead so your organization's data warehouse cost estimates will be more accurate.

\*These are the current default costs as of January 2025, and will be updated in Tobiko Cloud as warehouse cost defaults change.

> ## Where to find cost and savings estimates information

Reviewer 2: I suggest introducing savings before where to find them.

> Estimated costs and savings are displayed on the Tobiko Cloud homepage, production environment page, runs and plans pages, and individual model pages.

Reviewer 2: I believe this is the first time `estimated` is included. Should we add a paragraph to the first section with the disclaimer that it may not 100% match, but is based on the data retrieved from the datawarehouse itself? And then remove estimated from here and elsewhere.

Author: we do mention "estimated" beforehand but I will include a disclaimer so it's more clear

Cost information on each page will look similar to:

![Example of costs and savings data as seen on the Tobiko Cloud homepage](./costs_savings/costs-example.png)

## Savings Categories

> When calculating your data warehouse costs, Tobiko Cloud also estimates how much money it saved you!

Reviewer 2: Nit: who is `it` isn't clear here. It could be the dwh or tcloud.

> Cost savings are broken up into three main categories:

Reviewer 2: `There are three types of cost savings tracked by Tobiko Cloud:`

> - **Prevented Reruns**: If SQLMesh has already run an execution for a change in one environment, we won't need to rerun it in another environment (such as when a model change is backfilled on a development environment and then applied to prod).

Reviewer 2: `If SQLMesh has already executed the model for the specific interval(s) requested, it knows not to execute the interval(s) again (such as...)`

> - **Unaffected Downstream**: Because SQLMesh understands SQL, we know if a downstream model is not affected by an upstream change and can avoid re-execution.

Reviewer 2: `we can detect detect when a downstream model is not affected by an upstream change and avoid an re-execution that we know would produce the same data`

- **Virtual Environments**: Using the Virtual data environments feature means new environments can be created without backfilling models within the new environment.

### Where to find cost savings information

> Cost savings are shown in most of the places data warehouse costs are displayed. You can find how much Tobiko Cloud has saved you by viewing the homepage, production environment page, or individual model pages.

Reviewer 2: Do we not show on the plan page yet?

Author: Correct - I think we had wanted to be conservative with where we display aggregated costs, so those only live on the listed pages. The Plan page only has costs listed by execution in the model executions table, and we haven't added savings as a column there yet
